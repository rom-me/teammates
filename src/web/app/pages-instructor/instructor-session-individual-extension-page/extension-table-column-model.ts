import { InstructorPermissionRole } from '../../../types/api-request';

export interface StudentExtensionTableColumnModel {
  sectionName: string;
  teamName: string;
  name: string;
  email: string;
  extensionDeadline: number;
  hasExtension: boolean;
  isSelected: boolean;
}

export interface InstructorExtensionTableColumnModel {
  name: string;
  email: string;
  role?: InstructorPermissionRole;
  extensionDeadline: number;
  hasExtension: boolean;
  isSelected: boolean;
}
