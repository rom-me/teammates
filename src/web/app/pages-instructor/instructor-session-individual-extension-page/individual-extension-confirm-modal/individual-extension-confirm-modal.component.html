<fieldset [disabled]="isSubmitting">
  <div class="modal-header" [ngClass]="{'alert-primary': isExtendModal(), 'alert-danger' : isDeleteModal()}">
    <h5 class="modal-title" *ngIf="isExtendModal()">Confirm extension of feedback session</h5>
    <h5 class="modal-title" *ngIf="isDeleteModal()">Confirm deletion of feedback session extensions</h5>
    <button type="button" class="close" (click)="activeModal.dismiss()">
      <i class="fas fa-times"></i>
    </button>
  </div>

  <div class="modal-body">
    <div *ngIf="isExtendModal()">
      The new deadline of <b>{{ extensionTimestamp | formatDateDetail: feedbackSessionTimeZone }}</b> will be applied
      to the following <b>{{ selectedStudents.length }}</b> student(s) and <b>{{ selectedInstructors.length }}</b>
      instructor(s):
    </div>

    <div *ngIf="isDeleteModal()">
      The deadline of the following <b>{{ selectedStudents.length }}</b> student(s) and
      <b>{{ selectedInstructors.length }}</b> instructor(s) will be reverted back to the original deadline of
      <b>{{ extensionTimestamp | formatDateDetail: feedbackSessionTimeZone }}</b>:
    </div>

    <div class="table-responsive" *ngIf="selectedStudents.length !== 0">
      <h1>Students</h1>
      <table id="student-list-table" class="table table-bordered">
        <thead>
          <tr>
            <th class="sortable-header" (click)="sortStudentColumnsBy(SortBy.SECTION_NAME)">
              Section
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortStudentsBy === SortBy.SECTION_NAME && sortStudentOrder === SortOrder.DESC"
                  class="fas fa-sort-down"></i>
                <i *ngIf="sortStudentsBy === SortBy.SECTION_NAME && sortStudentOrder === SortOrder.ASC"
                  class="fas fa-sort-up"></i>
              </span>
            </th>
            <th class="sortable-header" (click)="sortStudentColumnsBy(SortBy.TEAM_NAME)">
              Team
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortStudentsBy === SortBy.TEAM_NAME && sortStudentOrder === SortOrder.DESC"
                  class="fas fa-sort-down"></i>
                <i *ngIf="sortStudentsBy === SortBy.TEAM_NAME && sortStudentOrder === SortOrder.ASC"
                  class="fas fa-sort-up"></i>
              </span>
            </th>
            <th class="sortable-header" (click)="sortStudentColumnsBy(SortBy.RESPONDENT_NAME)">
              Name
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortStudentsBy === SortBy.RESPONDENT_NAME && sortStudentOrder === SortOrder.DESC"
                  class="fas fa-sort-down"></i>
                <i *ngIf="sortStudentsBy === SortBy.RESPONDENT_NAME && sortStudentOrder === SortOrder.ASC"
                  class="fas fa-sort-up"></i>
              </span>
            </th>
            <th class="sortable-header" (click)="sortStudentColumnsBy(SortBy.RESPONDENT_EMAIL)">
              Email
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortStudentsBy === SortBy.RESPONDENT_EMAIL && sortStudentOrder === SortOrder.DESC"
                  class="fas fa-sort-down"></i>
                <i *ngIf="sortStudentsBy === SortBy.RESPONDENT_EMAIL && sortStudentOrder === SortOrder.ASC"
                  class="fas fa-sort-up"></i>
              </span>
            </th>
            <th class="sortable-header" (click)="sortStudentColumnsBy(SortBy.SESSION_END_DATE)">
              <span *ngIf="isExtendModal()">Original Deadline </span>
              <span *ngIf="isDeleteModal()">Current Deadline </span>
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortStudentsBy === SortBy.SESSION_END_DATE && sortStudentOrder === SortOrder.DESC"
                  class="fas fa-sort-down"></i>
                <i *ngIf="sortStudentsBy === SortBy.SESSION_END_DATE && sortStudentOrder === SortOrder.ASC"
                  class="fas fa-sort-up"></i>
              </span>
            </th>
          </tr>
        </thead>
        <tr *ngFor="let student of selectedStudents;">
          <td>{{ student.sectionName }}</td>
          <td>{{ student.teamName }}</td>
          <td>{{ student.name }}</td>
          <td>{{ student.email }}</td>
          <td>{{ student.extensionDeadline | formatDateDetail: feedbackSessionTimeZone }}</td>
        </tr>
      </table>
    </div>

    <div class="table-responsive" *ngIf="selectedInstructors.length !== 0">
      <h1>Instructors</h1>
      <table id="instructor-list-table" class="table table-bordered">
        <thead>
          <tr>
            <th class="sortable-header" (click)="sortInstructorsColumnsBy(SortBy.RESPONDENT_NAME)">
              Name
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortInstructorsBy === SortBy.RESPONDENT_NAME && sortInstructorOrder === SortOrder.DESC"
                  class="fas fa-sort-down"></i>
                <i *ngIf="sortInstructorsBy === SortBy.RESPONDENT_NAME && sortInstructorOrder === SortOrder.ASC"
                  class="fas fa-sort-up"></i>
              </span>
            </th>
            <th class="sortable-header" (click)="sortInstructorsColumnsBy(SortBy.RESPONDENT_EMAIL)">
              Email
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortInstructorsBy === SortBy.RESPONDENT_EMAIL && sortInstructorOrder === SortOrder.DESC"
                  class="fas fa-sort-down"></i>
                <i *ngIf="sortInstructorsBy === SortBy.RESPONDENT_EMAIL && sortInstructorOrder === SortOrder.ASC"
                  class="fas fa-sort-up"></i>
              </span>
            </th>
            <th class="sortable-header" (click)="sortInstructorsColumnsBy(SortBy.INSTRUCTOR_PERMISSION_ROLE)">
              Role
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortInstructorsBy === SortBy.INSTRUCTOR_PERMISSION_ROLE && sortInstructorOrder === SortOrder.DESC"
                  class="fas fa-sort-down"></i>
                <i *ngIf="sortInstructorsBy === SortBy.INSTRUCTOR_PERMISSION_ROLE && sortInstructorOrder === SortOrder.ASC"
                  class="fas fa-sort-up"></i>
              </span>
            </th>
            <th class="sortable-header" (click)="sortInstructorsColumnsBy(SortBy.SESSION_END_DATE)">
              <span *ngIf="isExtendModal()">Original Deadline </span>
              <span *ngIf="isDeleteModal()">Current Deadline </span>
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortInstructorsBy === SortBy.SESSION_END_DATE && sortInstructorOrder === SortOrder.DESC"
                  class="fas fa-sort-down"></i>
                <i *ngIf="sortInstructorsBy === SortBy.SESSION_END_DATE && sortInstructorOrder === SortOrder.ASC"
                  class="fas fa-sort-up"></i>
              </span>
            </th>
          </tr>
        </thead>
        <tr *ngFor="let instructor of selectedInstructors;">
          <td>{{ instructor.name }}</td>
          <td>{{ instructor.email }}</td>
          <td>{{ instructor.role | instructorRoleName }}</td>
          <td>{{ instructor.extensionDeadline | formatDateDetail: feedbackSessionTimeZone }}</td>
        </tr>
      </table>
    </div>

    <div class="form-check">
      <input class="form-check-input" type="checkbox" [(ngModel)]="isNotifyDeadlines" id="flexCheckChecked">
      <label class="form-check-label" for="flexCheckChecked" *ngIf="isExtendModal()">
        Notify affected students and/or instructors by email about this extension
      </label>
      <label class="form-check-label" for="flexCheckChecked" *ngIf="isDeleteModal()">
        Notify affected students and/or instructors by email about the cancellation of their extension
      </label>
    </div>

  </div>

  <br>

  <div class="modal-footer">
    <button type="button" class="btn btn-light" (click)="activeModal.dismiss()">No, cancel the Operation</button>
    <button *ngIf="isExtendModal()" type="button" class="btn btn-success modal-btn-ok"
      (click)="onConfirm()">Confirm</button>
    <button *ngIf="isDeleteModal()" type="button" class="btn btn-danger modal-btn-ok" (click)="onConfirm()">Delete
      Extension(s)</button>
  </div>
</fieldset>
