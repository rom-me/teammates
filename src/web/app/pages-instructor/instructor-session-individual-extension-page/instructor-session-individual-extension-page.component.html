<tm-loading-retry [shouldShowRetry]="hasLoadingFeedbackSessionFailed"
  [message]="'Failed to load course and feedback session'" (retryEvent)="loadFeedbackSessionAndIndividuals()">
  <div *tmIsLoading="isLoadingFeedbackSession">
    <div class="card bg-light top-padded text-center text-sm-left">
      <div class="card-body">
        <div class="row text-center">
          <div class="col-md-2 text-md-right font-bold">Course ID</div>
          <div id="time-zone" class="col-md-3 text-md-left">
            {{ courseId }}
          </div>
        </div>
        <div class="row text-center">
          <div class="col-md-2 text-md-right font-bold">Time Zone</div>
          <div id="time-zone" class="col-md-3 text-md-left">
            {{ feedbackSessionTimeZone }}
          </div>
        </div>
        <div class="row text-center">
          <div class="col-md-2 text-md-right font-bold">Course Name</div>
          <div id="course-name" class="col-md-10 text-md-left">
            {{ courseName }}
          </div>
        </div>
        <div class="row text-center">
          <div class="col-md-2 text-md-right font-bold">Session Name</div>
          <div id="course-name" class="col-md-10 text-md-left">
            {{ feedbackSessionName }}
          </div>
        </div>
        <div class="row text-center">
          <div class="col-md-2 text-md-right font-bold">Original Deadline</div>
          <div id="course-name" class="col-md-10 text-md-left">
            {{ feedbackSessionEndingTimestamp | formatDateDetail: feedbackSessionTimeZone }}
          </div>
        </div>
      </div>
    </div>
    <p> Select students or instructors from the lists and click the <b>Extend / Edit</b> or <b>Delete</b> buttons at
      the bottom of the page to update their feedback session deadline. </p>
  </div>
</tm-loading-retry>

<!-- TODO: Use lazy loading toggleable dropdowns-->
<tm-loading-retry [shouldShowRetry]="hasLoadedAllStudentsFailed" [message]="'Failed to load students'"
  (retryEvent)="loadFeedbackSessionAndIndividuals()">
  <div *tmIsLoading="isLoadingAllStudents">
    <h1>Students</h1>
    <div class="table-responsive">
      <table id="student-list-table" class="table table-bordered table-hover">
        <thead>
          <tr>
            <th (click)="selectAllStudents()">
              <label><input type="checkbox" id="select-all-student-btn" [checked]="isAllStudentsSelected" /> Select
                all</label>
            </th>
            <th class="sortable-header" (click)="sortStudentColumnsBy(SortBy.SECTION_NAME)">
              Section
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortStudentsBy === SortBy.SECTION_NAME && sortStudentOrder === SortOrder.DESC"
                  class="fas fa-sort-down"></i>
                <i *ngIf="sortStudentsBy === SortBy.SECTION_NAME && sortStudentOrder === SortOrder.ASC"
                  class="fas fa-sort-up"></i>
              </span>
            </th>
            <th class="sortable-header" (click)="sortStudentColumnsBy(SortBy.TEAM_NAME)">
              Team
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortStudentsBy === SortBy.TEAM_NAME && sortStudentOrder === SortOrder.DESC"
                  class="fas fa-sort-down"></i>
                <i *ngIf="sortStudentsBy === SortBy.TEAM_NAME && sortStudentOrder === SortOrder.ASC"
                  class="fas fa-sort-up"></i>
              </span>
            </th>
            <th class="sortable-header" (click)="sortStudentColumnsBy(SortBy.RESPONDENT_NAME)">
              Name
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortStudentsBy === SortBy.RESPONDENT_NAME && sortStudentOrder === SortOrder.DESC"
                  class="fas fa-sort-down"></i>
                <i *ngIf="sortStudentsBy === SortBy.RESPONDENT_NAME && sortStudentOrder === SortOrder.ASC"
                  class="fas fa-sort-up"></i>
              </span>
            </th>
            <th class="sortable-header" (click)="sortStudentColumnsBy(SortBy.RESPONDENT_EMAIL)">
              Email
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortStudentsBy === SortBy.RESPONDENT_EMAIL && sortStudentOrder === SortOrder.DESC"
                  class="fas fa-sort-down"></i>
                <i *ngIf="sortStudentsBy === SortBy.RESPONDENT_EMAIL && sortStudentOrder === SortOrder.ASC"
                  class="fas fa-sort-up"></i>
              </span>
            </th>
            <th class="sortable-header" (click)="sortStudentColumnsBy(SortBy.SESSION_END_DATE)">
              Current Deadline
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortStudentsBy === SortBy.SESSION_END_DATE && sortStudentOrder === SortOrder.DESC"
                  class="fas fa-sort-down"></i>
                <i *ngIf="sortStudentsBy === SortBy.SESSION_END_DATE && sortStudentOrder === SortOrder.ASC"
                  class="fas fa-sort-up"></i>
              </span>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr *ngFor="let student of studentsOfCourse; index as i" id="student-row-{{ i }}" (click)="selectStudent(i)"
            [ngClass]="{ 'table-warning': student.hasExtension === true }">
            <td class="align-center">
              <label><input type="checkbox" [checked]="student.isSelected" /></label>
            </td>
            <td>{{ student.sectionName }}</td>
            <td>{{ student.teamName }}</td>
            <td>{{ student.name }}</td>
            <td>{{ student.email }}</td>
            <td>{{ student.extensionDeadline | formatDateDetail: feedbackSessionTimeZone }}</td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</tm-loading-retry>

<br>

<tm-loading-retry [shouldShowRetry]="hasLoadedAllInstructorsFailed" [message]="'Failed to load instructors'"
  (retryEvent)="loadFeedbackSessionAndIndividuals()">
  <div *tmIsLoading="isLoadingAllInstructors">
    <h1>Instructors</h1>
    <div class="table-responsive">
      <table id="instructor-list-table" class="table table-bordered table-hover">
        <thead>
          <tr>
            <th (click)="selectAllInstructors()">
              <label><input type="checkbox" id="select-all-instructor-btn" [checked]="isAllInstructorsSelected" />
                Select all</label>
            </th>
            <th class="sortable-header" (click)="sortInstructorsColumnsBy(SortBy.RESPONDENT_NAME)">
              Name
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortInstructorsBy === SortBy.RESPONDENT_NAME && sortInstructorOrder === SortOrder.DESC"
                  class="fas fa-sort-down"></i>
                <i *ngIf="sortInstructorsBy === SortBy.RESPONDENT_NAME && sortInstructorOrder === SortOrder.ASC"
                  class="fas fa-sort-up"></i>
              </span>
            </th>
            <th class="sortable-header" (click)="sortInstructorsColumnsBy(SortBy.RESPONDENT_EMAIL)">
              Email
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortInstructorsBy === SortBy.RESPONDENT_EMAIL && sortInstructorOrder === SortOrder.DESC"
                  class="fas fa-sort-down"></i>
                <i *ngIf="sortInstructorsBy === SortBy.RESPONDENT_EMAIL && sortInstructorOrder === SortOrder.ASC"
                  class="fas fa-sort-up"></i>
              </span>
            </th>
            <th class="sortable-header" (click)="sortInstructorsColumnsBy(SortBy.INSTRUCTOR_PERMISSION_ROLE)">
              Role
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortInstructorsBy === SortBy.INSTRUCTOR_PERMISSION_ROLE && sortInstructorOrder === SortOrder.DESC"
                  class="fas fa-sort-down"></i>
                <i *ngIf="sortInstructorsBy === SortBy.INSTRUCTOR_PERMISSION_ROLE && sortInstructorOrder === SortOrder.ASC"
                  class="fas fa-sort-up"></i>
              </span>
            </th>
            <th class="sortable-header" (click)="sortInstructorsColumnsBy(SortBy.SESSION_END_DATE)">
              Current Deadline
              <span class="fa-stack">
                <i class="fas fa-sort"></i>
                <i *ngIf="sortInstructorsBy === SortBy.SESSION_END_DATE && sortInstructorOrder === SortOrder.DESC"
                  class="fas fa-sort-down"></i>
                <i *ngIf="sortInstructorsBy === SortBy.SESSION_END_DATE && sortInstructorOrder === SortOrder.ASC"
                  class="fas fa-sort-up"></i>
              </span>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr *ngFor="let instructor of instructorsOfCourse; index as i" id="instructor-row-{{ i }}" (click)="selectIntructor(i)"
            [ngClass]="{ 'table-warning': instructor.hasExtension === true }">
            <td class="align-center">
              <label><input type="checkbox" [checked]="instructor.isSelected" /></label>
            </td>
            <td>{{ instructor.name }}</td>
            <td>{{ instructor.email }}</td>
            <td>{{ instructor.role | instructorRoleName }}</td>
            <td>{{ instructor.extensionDeadline | formatDateDetail: feedbackSessionTimeZone }}</td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</tm-loading-retry>
<br>
<div class="align-row-right">
  <div>
    <button type="button" id="extend-btn" class="btn btn-success" (click)="onExtend()"
      [disabled]="!hasSelected()">Extend / Edit</button>
  </div>
  <div>
    <button type="button" id="delete-btn" class="btn btn-danger" (click)="onDelete()"
      [disabled]="!hasSelectedValidForDeletion()">Delete</button>
  </div>
</div>
