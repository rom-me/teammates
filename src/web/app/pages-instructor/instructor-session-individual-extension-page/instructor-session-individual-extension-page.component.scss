.align-row-right {
  justify-content: right;
  display: flex;
  flex-direction: row;
  gap: 20px;
}
