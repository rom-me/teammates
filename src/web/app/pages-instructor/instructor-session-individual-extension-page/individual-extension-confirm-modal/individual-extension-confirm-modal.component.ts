import { Component, EventEmitter, Input, Output } from '@angular/core';
import { NgbActiveModal } from '@ng-bootstrap/ng-bootstrap';
import { TableComparatorService } from '../../../../services/table-comparator.service';
import { SortBy, SortOrder } from '../../../../types/sort-properties';
import { InstructorExtensionTableColumnModel, StudentExtensionTableColumnModel } from '../extension-table-column-model';

export enum ExtensionModalType {
  EXTEND,
  DELETE,
}

@Component({
  selector: 'tm-individual-extension-confirm-modal',
  templateUrl: './individual-extension-confirm-modal.component.html',
  styleUrls: ['./individual-extension-confirm-modal.component.scss'],
})
export class IndividualExtensionConfirmModalComponent {
  @Input()
  modalType: ExtensionModalType = ExtensionModalType.EXTEND;

  @Input()
  selectedStudents: StudentExtensionTableColumnModel[] = [];

  @Input()
  selectedInstructors: InstructorExtensionTableColumnModel[] = [];

  @Input()
  extensionTimestamp: number = 0;

  @Input()
  feedbackSessionTimeZone: string = '';

  @Output()
  onConfirmExtensionCallBack: EventEmitter<boolean> = new EventEmitter();

  constructor(public activeModal: NgbActiveModal, private tableComparatorService: TableComparatorService) {}

  SortBy: typeof SortBy = SortBy;
  SortOrder: typeof SortOrder = SortOrder;
  sortStudentsBy: SortBy = SortBy.SESSION_END_DATE;
  sortStudentOrder: SortOrder = SortOrder.DESC;
  sortInstructorsBy: SortBy = SortBy.SESSION_END_DATE;
  sortInstructorOrder: SortOrder = SortOrder.DESC;

  isSubmitting: boolean = false;
  isNotifyDeadlines: boolean = false;

  onConfirm(): void {
    this.isSubmitting = true;
    this.onConfirmExtensionCallBack.emit(this.isNotifyDeadlines);
  }

  isDeleteModal(): boolean {
    return this.modalType === ExtensionModalType.DELETE;
  }

  isExtendModal(): boolean {
    return this.modalType === ExtensionModalType.EXTEND;
  }

  sortStudentColumnsBy(by: SortBy): void {
    this.sortStudentsBy = by;
    this.sortStudentOrder = this.sortStudentOrder === SortOrder.DESC ? SortOrder.ASC : SortOrder.DESC;
    this.selectedStudents.sort(this.sortStudentPanelsBy(by));
  }

  sortStudentPanelsBy(
    by: SortBy,
  ): (a: StudentExtensionTableColumnModel, b: StudentExtensionTableColumnModel) => number {
    return (a: StudentExtensionTableColumnModel, b: StudentExtensionTableColumnModel): number => {
      let strA: string;
      let strB: string;
      switch (by) {
        case SortBy.SECTION_NAME:
          strA = a.sectionName;
          strB = b.sectionName;
          break;
        case SortBy.TEAM_NAME:
          strA = a.sectionName;
          strB = b.sectionName;
          break;
        case SortBy.RESPONDENT_NAME:
          strA = a.name;
          strB = b.name;
          break;
        case SortBy.RESPONDENT_EMAIL:
          strA = a.email;
          strB = b.email;
          break;
        case SortBy.SESSION_END_DATE:
          strA = a.extensionDeadline.toString();
          strB = b.extensionDeadline.toString();
          break;
        default:
          strA = '';
          strB = '';
      }
      return this.tableComparatorService.compare(by, this.sortStudentOrder, strA, strB);
    };
  }

  sortInstructorsColumnsBy(by: SortBy): void {
    this.sortInstructorsBy = by;
    this.sortInstructorOrder = this.sortInstructorOrder === SortOrder.DESC ? SortOrder.ASC : SortOrder.DESC;
    this.selectedInstructors.sort(this.sortInstructorPanelsBy(by));
  }

  sortInstructorPanelsBy(
    by: SortBy,
  ): (a: InstructorExtensionTableColumnModel, b: InstructorExtensionTableColumnModel) => number {
    return (a: InstructorExtensionTableColumnModel, b: InstructorExtensionTableColumnModel): number => {
      let strA: string;
      let strB: string;
      switch (by) {
        case SortBy.RESPONDENT_NAME:
          strA = a.name;
          strB = b.name;
          break;
        case SortBy.RESPONDENT_EMAIL:
          strA = a.email;
          strB = b.email;
          break;
        case SortBy.INSTRUCTOR_PERMISSION_ROLE:
          strA = a.role || '';
          strB = b.role || '';
          break;
        case SortBy.SESSION_END_DATE:
          strA = a.extensionDeadline.toString();
          strB = b.extensionDeadline.toString();
          break;
        default:
          strA = '';
          strB = '';
      }
      return this.tableComparatorService.compare(by, this.sortInstructorOrder, strA, strB);
    };
  }
}
