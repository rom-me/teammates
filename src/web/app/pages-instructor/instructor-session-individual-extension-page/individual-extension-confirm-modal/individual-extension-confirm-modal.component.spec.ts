import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { NgbActiveModal } from '@ng-bootstrap/ng-bootstrap';
import { TimezoneService } from '../../../../services/timezone.service';
import {
  FeedbackSession,
  FeedbackSessionPublishStatus,
  FeedbackSessionSubmissionStatus,
  ResponseVisibleSetting,
  SessionVisibleSetting,
} from '../../../../types/api-output';
import { StudentExtensionTableColumnModel } from '../extension-table-column-model';
import { InstructorSessionIndividualExtensionPageModule } from '../instructor-session-individual-extension-page.module';
import { IndividualExtensionConfirmModalComponent } from './individual-extension-confirm-modal.component';

describe('IndividualExtensionConfirmModalComponent', () => {
  const testFeedbackSession: FeedbackSession = {
    courseId: 'testId1',
    timeZone: 'Asia/Singapore',
    feedbackSessionName: 'Test Session',
    instructions: 'Instructions',
    submissionStartTimestamp: 1000000000000,
    submissionEndTimestamp: 1500000000000,
    gracePeriod: 0,
    sessionVisibleSetting: SessionVisibleSetting.AT_OPEN,
    responseVisibleSetting: ResponseVisibleSetting.AT_VISIBLE,
    submissionStatus: FeedbackSessionSubmissionStatus.OPEN,
    publishStatus: FeedbackSessionPublishStatus.PUBLISHED,
    isClosingEmailEnabled: true,
    isPublishedEmailEnabled: true,
    createdAtTimestamp: 0,
    studentDeadlines: {},
    instructorDeadlines: {},
  };

  const studentModel1: StudentExtensionTableColumnModel = {
    sectionName: 'Test Section 1',
    teamName: 'Test Section 1',
    name: 'Test Student 1',
    email: 'testStudent1@gmail.com',
    extensionDeadline: 1500000000000,
    hasExtension: false,
    isSelected: true,
  };
  const studentModel2: StudentExtensionTableColumnModel = {
    sectionName: 'Test Section 2',
    teamName: 'Test Section 2',
    name: 'Test Student 2',
    email: 'testStudent2@gmail.com',
    extensionDeadline: 1510000000000,
    hasExtension: true,
    isSelected: true,
  };
  const studentModel3: StudentExtensionTableColumnModel = {
    sectionName: 'Test Section 3',
    teamName: 'Test Section 3',
    name: 'Test Student 3',
    email: 'testStudent3@gmail.com',
    extensionDeadline: 1510000000000,
    hasExtension: true,
    isSelected: true,
  };

  const testTimeString = '5 Apr 2000 2:00:00';

  let component: IndividualExtensionConfirmModalComponent;
  let fixture: ComponentFixture<IndividualExtensionConfirmModalComponent>;
  let timeZoneService: TimezoneService;

  beforeEach(
    waitForAsync(() => {
      TestBed.configureTestingModule({
        imports: [HttpClientTestingModule, InstructorSessionIndividualExtensionPageModule],
        providers: [NgbActiveModal],
      }).compileComponents();
    }),
  );

  beforeEach(() => {
    fixture = TestBed.createComponent(IndividualExtensionConfirmModalComponent);
    timeZoneService = TestBed.inject(TimezoneService);
    jest.spyOn(timeZoneService, 'formatToString').mockReturnValue(testTimeString);
    component = fixture.componentInstance;
    component.feedbackSessionTimeZone = 'Asia/Singapore';
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with the extended students and instructors', () => {
    component.selectedStudents = [studentModel1, studentModel2, studentModel3];
    component.extensionTimestamp = testFeedbackSession.submissionEndTimestamp;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
