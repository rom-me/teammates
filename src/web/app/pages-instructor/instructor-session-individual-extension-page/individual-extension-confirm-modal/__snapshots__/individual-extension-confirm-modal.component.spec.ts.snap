// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndividualExtensionConfirmModalComponent should snap with the extended students and instructors 1`] = `
<tm-individual-extension-confirm-modal
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  activeModal={[Function NgbActiveModal]}
  extensionTimestamp={[Function Number]}
  feedbackSessionTimeZone={[Function String]}
  isNotifyDeadlines="false"
  isSubmitting="false"
  modalType="0"
  onConfirmExtensionCallBack={[Function EventEmitter_]}
  selectedInstructors={[Function Array]}
  selectedStudents={[Function Array]}
  sortInstructorOrder="0"
  sortInstructorsBy={[Function Number]}
  sortStudentOrder="0"
  sortStudentsBy={[Function Number]}
  tableComparatorService={[Function TableComparatorService]}
>
  <fieldset>
    <div
      class="modal-header alert-primary"
    >
      <h5
        class="modal-title"
      >
        Confirm extension of feedback session
      </h5>
      <button
        class="close"
        type="button"
      >
        <i
          class="fas fa-times"
        />
      </button>
    </div>
    <div
      class="modal-body"
    >
      <div>
         The new deadline of 
        <b>
          5 Apr 2000 2:00:00
        </b>
         will be applied to the following 
        <b>
          3
        </b>
         student(s) and 
        <b>
          0
        </b>
         instructor(s): 
      </div>
      <div
        class="table-responsive"
      >
        <h1>
          Students
        </h1>
        <table
          class="table table-bordered"
          id="student-list-table"
        >
          <thead>
            <tr>
              <th
                class="sortable-header"
              >
                 Section 
                <span
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </th>
              <th
                class="sortable-header"
              >
                 Team 
                <span
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </th>
              <th
                class="sortable-header"
              >
                 Name 
                <span
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </th>
              <th
                class="sortable-header"
              >
                 Email 
                <span
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                </span>
              </th>
              <th
                class="sortable-header"
              >
                <span>
                  Original Deadline 
                </span>
                <span
                  class="fa-stack"
                >
                  <i
                    class="fas fa-sort"
                  />
                  <i
                    class="fas fa-sort-down"
                  />
                </span>
              </th>
            </tr>
          </thead>
          <tr>
            <td>
              Test Section 1
            </td>
            <td>
              Test Section 1
            </td>
            <td>
              Test Student 1
            </td>
            <td>
              testStudent1@gmail.com
            </td>
            <td>
              5 Apr 2000 2:00:00
            </td>
          </tr>
          <tr>
            <td>
              Test Section 2
            </td>
            <td>
              Test Section 2
            </td>
            <td>
              Test Student 2
            </td>
            <td>
              testStudent2@gmail.com
            </td>
            <td>
              5 Apr 2000 2:00:00
            </td>
          </tr>
          <tr>
            <td>
              Test Section 3
            </td>
            <td>
              Test Section 3
            </td>
            <td>
              Test Student 3
            </td>
            <td>
              testStudent3@gmail.com
            </td>
            <td>
              5 Apr 2000 2:00:00
            </td>
          </tr>
        </table>
      </div>
      <div
        class="form-check"
      >
        <input
          class="form-check-input ng-untouched ng-pristine ng-valid"
          id="flexCheckChecked"
          type="checkbox"
        />
        <label
          class="form-check-label"
          for="flexCheckChecked"
        >
           Notify affected students and/or instructors by email about this extension 
        </label>
      </div>
    </div>
    <br />
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-light"
        type="button"
      >
        No, cancel the Operation
      </button>
      <button
        class="btn btn-success modal-btn-ok"
        type="button"
      >
        Confirm
      </button>
    </div>
  </fieldset>
</tm-individual-extension-confirm-modal>
`;
